src/validators/is_instance.rs

```diff
@@ -66,27 +59,27 @@ impl Validator for IsInstanceValidator {
         &'s self,
         py: Python<'data>,
         input: &'data impl Input<'data>,
-        _extra: &Extra,
+        extra: &Extra,
         _definitions: &'data Definitions<CombinedValidator>,
         _recursion_guard: &'s mut RecursionGuard,
     ) -> ValResult<'data, PyObject> {
-        match input.input_is_instance(self.class.as_ref(py), self.json_types)? {
-            true => {
-                if input.get_type().is_json() {
-                    if let Some(ref json_function) = self.json_function {
-                        return json_function
-                            .call1(py, (input.to_object(py),))
-                            .map_err(|e| convert_err(py, e, input));
-                    }
+        match extra.mode {
+            InputType::Json => Err(ValError::InternalErr(PyNotImplementedError::new_err(
+                "Cannot check isinstance when validating from json,\
+                            use a JsonOrPython validator instead.",
+            ))),
```

Comment:
Happy to do something else here. We, unfortunately, have to do this at runtime because we build a single validator for all of the options. Since we have `StrictOrLax` schema and `JsonOrPython` we may want to eventually re-evaluate that choice which would let us do optimizations like inlining those two branches and doing this check statically instead of during validation, which would lead to a better user experience.